### PR TITLE
reconcile oauth2client

### PR DESCRIPTION
**Pull request by auhlig**
I was doing that manually before, but this should be implemented and automated instead.

### CODE REVIEW

**Comment by auhlig** [comment]
[This error](https://github.com/cloudoperators/greenhouse/actions/runs/8815917330/job/24198862111?pr=160#step:4:567) seems unrelated, or? @IvoGoman?

**Comment by IvoGoman** [comment]
> [This error](https://github.com/cloudoperators/greenhouse/actions/runs/8815917330/job/24198862111?pr=160#step:4:567) seems unrelated, or? @IvoGoman?

Yes, 

> [This error](https://github.com/cloudoperators/greenhouse/actions/runs/8815917330/job/24198862111?pr=160#step:4:567) seems unrelated, or? @IvoGoman?

Yes, let's keep an eye on it. If it keeps being flaky this needs to be addressed.

File: pkg/controllers/organization/dex_controller.go [hunk 1]
```diff
@@ -166,3 +212,18 @@ func ensureCallbackURL(url string) string {
...
+
+func encodedOAuth2ClientNameForOrganization(orgName string) string {
+	// See https://github.com/dexidp/dex/issues/1606 for encoding.
+	return strings.TrimRight(base32.
+		NewEncoding("abcdefghijklmnopqrstuvwxyz234567").
+		EncodeToString(fnv.New64().Sum([]byte(orgName))), "=",
+	)
+}
+
+func appendStringToSliceIfNotContains(theString string, theSlice []string) []string {
```
[the remark this answers is not in the document]
Comment (auhlig) [reply pkg/controllers/organization/dex_controller.go:224]: I remember doing something similar before, but cannot find it.
The function is now moved to the util package and made public.

File: pkg/controllers/organization/dex_controller.go [hunk 2]
```diff
@@ -166,3 +212,18 @@ func ensureCallbackURL(url string) string {
 	}
 	return url
 }
+
+func encodedOAuth2ClientNameForOrganization(orgName string) string {
```
[the remark this answers is not in the document]
Comment (auhlig) [reply pkg/controllers/organization/dex_controller.go]: Sure. It's shorter now
I'm usually a fan of longer but explicit function names

File: pkg/util/slices.go [hunk 3]
```diff
@@ -1,3 +1,6 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
```
Comment (auhlig) [inline pkg/util/slices.go:1]: why didn't the license check action work?

File: pkg/controllers/organization/dex_controller.go [hunk 4]
```diff
@@ -155,6 +165,42 @@ func (r *DexReconciler) discoverOIDCRedirectURL(ctx context.Context, org *greenh
...
+	var oAuth2Client = new(dexapi.OAuth2Client)
+	oAuth2Client.ObjectMeta.Name = encodedOAuth2ClientNameForOrganization(org.Name)
+	oAuth2Client.ObjectMeta.Namespace = r.Namespace
+
+	result, err := clientutil.CreateOrPatch(ctx, r.Client, oAuth2Client, func() error {
+		oAuth2Client.Client.Public = true
+		oAuth2Client.Client.ID = org.Name
+		oAuth2Client.Client.Name = org.Name
+		if oAuth2Client.RedirectURIs == nil {
+			oAuth2Client.RedirectURIs = make([]string, 0)
```
[the remark this answers is not in the document]
Comment (auhlig) [reply pkg/controllers/organization/dex_controller.go]: I know what will be overlooked once another redirect URL is added :)

File: pkg/util/slices.go (hunk 3 above)
Comment (IvoGoman) [reply pkg/util/slices.go:1]: https://github.com/cloudoperators/greenhouse/actions/runs/8823560471/job/24224268387

**Review by IvoGoman — APPROVED** [review]

**Comment by auhlig** [comment]
rebased on main

**Pull request merged**